### PR TITLE
image name must not have leading or trailing whitespace

### LINKED ISSUE #47490: Pod container image is allowed to just be a space character
**Is this a BUG REPORT or FEATURE REQUEST?** 
BUG

**Kubernetes version** (use `kubectl version`):
kubernetes 1.5+

**Environment**:
n/a

**What happened**:
examining events that happened in my cluster, i see an event like the following:

```
xyz   5s        7d        46927     mongoxyz   Pod       spec.containers{mongodb}                  Warning   InspectFailed   {kubelet ip-removed}   Failed to apply default image tag " ": couldn't parse image reference " ": invalid reference format
```

this surprised me, as it implies a container had an imageName=" "

**What you expected to happen**:
i expected a validation failure error if a pod.spec.container[i].imageName had leading or trailing whitespace.

**How to reproduce it** (as minimally and precisely as possible):
kubectl create -f my-pod
kubectl edit pod/my-pod -o json
change image to " "
save

**Anything else we need to know**:
nope

### PR DESCRIPTION
**Pull request by derekwaynecarr**
**What this PR does / why we need it**:
verifies that an image name can not have leading or trailing whitespace

**Which issue this PR fixes**
fixes https://github.com/kubernetes/kubernetes/issues/47490

**Special notes for your reviewer**:
i was surprised we had not caught this, so if there is an image spec that says leading and trailing whitespace is a good thing, i am open to correction.

i was made aware of downstream users of validate pod template spec that used " " as a special token.  as a result, i only do the validation of image name " " in the `Pod` only.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
@eparis @smarterclayton -- in debugging events on our cluster, i observed 3 pods had an image name set to " ", and this resulted in ImageFailed events happening about 47k times for this pod over ~7d.

File: pkg/api/validation/validation_test.go [hunk 1]
```diff
@@ -3233,9 +3233,10 @@ func TestValidateContainers(t *testing.T) {
 		AllowPrivileged: false,
 	})
 	errorCases := map[string][]api.Container{
-		"zero-length name":     {{Name: "", Image: "image", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: "File"}},
-		"name > 63 characters": {{Name: strings.Repeat("a", 64), Image: "image", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: "File"}},
-		"name not a DNS label": {{Name: "a.b.c", Image: "image", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: "File"}},
+		"zero-length name":                      {{Name: "", Image: "image", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: "File"}},
+		"image leading and trailing whitespace": {{Name: "", Image: "  image   ", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: "File"}},
```
Comment (gyliu513) [inline pkg/api/validation/validation_test.go]: How about add another case of enabling image = "   " like what you did in the test?

**Comment by derekwaynecarr** [comment]
@gyliu513 - done

**Comment by smarterclayton** [comment]
/approve

This can't break anyone (docker daemon won't pull), it's confusing and accidental, and it causes issues for the rest of the cluster

**Review by gyliu513 — COMMENTED** [review]
/lgtm

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1975,6 +1975,9 @@ func validateContainers(containers []api.Container, volumes sets.String, fldPath
 		if len(ctr.Image) == 0 {
 			allErrs = append(allErrs, field.Required(idxPath.Child("image"), ""))
 		}
+		if len(strings.TrimSpace(ctr.Image)) != len(ctr.Image) {
```
Comment (liggitt) [inline pkg/api/validation/validation.go]: is this used to validate all objects with pod specs, or just pods?

Comment (smarterclayton) [inline pkg/api/validation/validation.go]: Yeah, can only be final pods.

Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: right, anyone that calls ValidatePodSpec, so this is picked up by all the job/rc/...

Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: i don't see how image is special.  we validate every other aspect of the pod spec when doing a validate pod template (volumes, affinity, etc.)

**Comment by smarterclayton** [comment]
My concern is that for everything that uses a pod spec template today, `"image": " "` is valid JSON.  In practice, you'd need an admission controller to leverage that, so it's unlikely to be a serious hole for anyone.

**Comment by derekwaynecarr** [comment]
@smarterclayton - opened https://github.com/openshift/origin/issues/14659

i will update this pr and reference that issue for rationale until we transitioned.

**Comment by derekwaynecarr** [comment]
@smarterclayton @liggitt - updated per request.

**Comment by smarterclayton** [comment]
/lgtm

with the restricted scope

**Pull request merged**